src/data/sym/basic.lean

```diff
@@ -150,4 +155,100 @@ instance inhabited_sym' [inhabited α] (n : ℕ) : inhabited (sym' α n) :=
+instance has_emptyc : has_emptyc (sym α 0) := ⟨0⟩
+
+instance subsingleton (n : ℕ) [subsingleton α] : subsingleton (sym α n) :=
+⟨begin
+  rintros ⟨a, ha⟩ ⟨b, hb⟩,
+  simp only [subtype.mk.inj_eq],
+  induction a using multiset.case_strong_induction_on with k hk ih generalizing n b,
+  { rw [ha.symm, multiset.card_zero] at hb, exact (multiset.card_eq_zero.mp hb).symm, },
+  { cases n,
+    { exact (multiset.cons_ne_zero (multiset.card_eq_zero.mp ha)).elim },
+    { by_cases hzero : b = 0,
+      { rw [hzero, multiset.card_zero] at hb,
+        exact (n.succ_ne_zero hb.symm).elim },
+      { have hmem := multiset.exists_mem_of_ne_zero hzero,
+        rcases hmem with ⟨r, hr⟩,
+        cases multiset.exists_cons_of_mem hr,
+        rw h,
+        have heq := @ih hk rfl.ge n w begin
+          rw multiset.card_cons at ha,
+          refine nat.succ.inj ha,
+        end begin
+          rw [h, multiset.card_cons] at hb,
+          refine nat.succ.inj hb,
+        end,
+        rw [heq, subsingleton.elim k r] } } }
+end⟩
+
+instance unique (n : ℕ) [unique α] : unique (sym α n) := unique.mk' _
+
+instance is_empty (n : ℕ) [is_empty α] : is_empty (sym α n.succ) :=
+⟨begin
+  intro h,
+  rw sym at h,
+  refine is_empty.exists_iff.mp (@multiset.exists_mem_of_ne_zero _ h.val _),
+  intro y,
+  have z := h.property,
+  rw [y, multiset.card_zero] at z,
+  exact (nat.succ_ne_zero n z.symm).elim,
+end⟩
+
+/-- `repeat a n` is the sym containing only `a` with multiplicity `n`. -/
+def repeat (a : α) (n : ℕ) : sym α n := ⟨multiset.repeat a n, multiset.card_repeat _ _⟩
+
+lemma repeat_left_injective (n : ℕ) (h : n ≠ 0) : function.injective (λ x : α, repeat x n) :=
+begin
+  intros a b x,
+  simp only [repeat, subtype.mk.inj_eq] at x,
+  exact (multiset.repeat_left_inj a b n h).mp x,
+end
+
+lemma repeat_left_inj (a b : α) (n : ℕ) (h : n ≠ 0) : repeat a n = repeat b n ↔ a = b :=
+(repeat_left_injective n h).eq_iff
+
+instance nontrivial (n : ℕ) [nontrivial α] : nontrivial (sym α (n + 1)) :=
+(repeat_left_injective n.succ n.succ_ne_zero).nontrivial
+
+def map {α β : Type*} {n : ℕ} (f : α → β) (x : sym α n) : sym β n :=
+⟨x.val.map f, by simpa [multiset.card_map] using x.property⟩
+
+@[simp] lemma mem_map {α β : Type*} {n : ℕ} {f : α → β} {b : β} {l : sym α n} :
+  b ∈ sym.map f l ↔ ∃ a, a ∈ l ∧ f a = b := multiset.mem_map
+
+@[simp] lemma map_id {α : Type*} {n : ℕ} (s : sym α n) : sym.map id s = s :=
+by simp [sym.map, subtype.mk.inj_eq]
+
+@[simp] lemma map_map {α β γ : Type*} {n : ℕ} (g : β → γ) (f : α → β) (s : sym α n) :
+  sym.map g (sym.map f s) = sym.map (g ∘ f) s :=
+by simp [sym.map, subtype.mk.inj_eq]
+
+@[simp] lemma map_zero {α β : Type*} (f : α → β) :
+  sym.map f (0 : sym α 0) = (0 : sym β 0) :=
+begin
+  rw sym.has_zero,
+  simp only [sym.map, multiset.map_zero],
+  rw sym.has_zero,
+  simp only [subtype.mk_eq_mk],
+end
+
+@[simp] lemma map_cons {α β : Type*} {n : ℕ} (f : α → β) (a : α) (s : sym α n) :
+  sym.map f (a::s) = (f a)::sym.map f s :=
```

Comment:
Dot notation makes this a little more readable:
```suggestion
  (a :: s).map f = f a :: s.map f :=
```
Note that the meaning is identical to lean